### PR TITLE
Redelegation integration tests

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
Closes https://github.com/cosmos/interchain-security/issues/255, also adds a corresponding e2e test

### CODE REVIEW

**Review by jtremback — APPROVED** [review]
Nice. You got some e2e tests for the functionality in there too?

**Comment by smarshall-spitzbart** [comment]
> Nice. You got some e2e tests for the functionality in there too?

Yea I added a single ```TestRedelegationNoConsumer``` e2e test. The equivalent e2e test with a consumer shouldn't be crazy hard to implement, but I left it as a TODO for now

File: e2e-tests/valset_update_test.go [hunk 1]
```diff
@@ -124,3 +125,71 @@ func (suite *ConsumerKeeperTestSuite) TestUnbondMaturePackets() {
...
+	providerUnbondingPeriod := stakingKeeper.UnbondingTime(s.providerCtx())
+	s.Require().Equal(
+		s.providerCtx().BlockTime().Add(providerUnbondingPeriod),
+		redelgation.Entries[0].CompletionTime)
+
+	// Check validator powers
+	val1PowerAfter := stakingKeeper.GetLastValidatorPower(s.providerCtx(), val1Addr)
+	s.Require().Greater(val1PowerAfter, val1PowerBefore)
+	s.Require().Equal(val0PowerBefore, stakingKeeper.GetLastValidatorPower(s.providerCtx(), val0Addr))
+}
```
Comment (mpoke) [inline e2e-tests/valset_update_test.go:190]: `TestRedelegationNoConsumer` should also check that after the unbonding period passes on the provider, the redelegation entry is no longer there. Use `TestUnbondingNoConsumer` as an example.

File: e2e-tests/valset_update_test.go [hunk 2]
```diff
@@ -124,3 +125,71 @@ func (suite *ConsumerKeeperTestSuite) TestUnbondMaturePackets() {
...
+
+// TestRedelegationWithConsumer tests a redelegate transaction submitted on a provider chain with a consumer
+func (s *ProviderTestSuite) TestRedelegationWithConsumer() {
+	// TODO
```
Comment (mpoke) [inline e2e-tests/valset_update_test.go:194]: Use `TestUndelegationProviderFirst` or `TestUndelegationConsumerFirst` as examples.

**Review by mpoke — CHANGES_REQUESTED** [review]
The integration tests LGTM. Some comments below re. the e2e tests.

File: e2e-tests/common_test.go [hunk 3]
```diff
@@ -108,6 +108,21 @@ func undelegate(s *ProviderTestSuite, delAddr sdk.AccAddress, valAddr sdk.ValAdd
...
+	completionTime, err := s.providerChain.App.(*appProvider.App).StakingKeeper.BeginRedelegation(
+		s.providerCtx(),
+		delAddr,
+		valSrcAddr,
+		ValDstAddr,
+		sharesAmount,
+	)
+	s.Require().NoError(err)
+	providerUnbondingPeriod := s.providerChain.App.GetStakingKeeper().UnbondingTime(s.providerCtx())
+	s.Require().Equal(s.providerCtx().BlockHeader().Time.Add(providerUnbondingPeriod), completionTime)
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply e2e-tests/common_test.go]: Good catch, I've added a case that validator being UNBONDING as well

File: e2e-tests/valset_update_test.go [hunk 4]
```diff
@@ -124,3 +125,71 @@ func (suite *ConsumerKeeperTestSuite) TestUnbondMaturePackets() {
 	suite.Require().Equal(uint64(1), commitments[0].Sequence, "did not send VSCMatured packet for VSC packet 1")
 	suite.Require().Equal(uint64(2), commitments[1].Sequence, "did not send VSCMatured packet for VSC packet 2")
 }
+
+// TestRedelegationNoConsumer tests a redelegate transaction submitted on a provider chain with no consumers
+func (s *ProviderTestSuite) TestRedelegationNoConsumer() {
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply e2e-tests/valset_update_test.go]: Agreed, I've moved these tests. However ```valset_update_test.go``` is pretty related to unbonding, and this makes the distinction between those two files difficult to understand. Anytime there is an unbonding operation, there will be a valset update. 

Maybe we consider consolidating and/or renaming the two files?

[the remark this answers is not in the document]
Comment (mpoke) [reply e2e-tests/valset_update_test.go]: Validator Set Update is the name of the entire subprotocol (see https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/methods.md#validator-set-update). Unbonding it's just a part of it that includes also the modifications done to the Staking module (see https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/methods.md#validator-set-update). All the files in e2e-tests can be named after the subprotocol that is being tested.

**Pull request merged**